src/libexec/zmmtastatus

```diff
@@ -49,8 +50,7 @@ sub mtaIsRunning {
     chomp $pid;
     if ($pid ne "") {
       print "MTA process $pid is "  if $verbose;
-      system("kill -0 $pid 2> /dev/null");
-      if ($? == 0) {
+      if (readlink("/proc/$pid/exe") eq $postfixMaster) {
```

Comment:
How about using a systemd unit for the MTA in the future rather extending the current whole script enchilada?

Reply:
Why not just let postfix do it?... You are calling postfix to start/stop currently. They have a lock on that master.pid file. That would also future proof the solution should postfix change in the future.  Once I saw how they were handling postfix start and checking for that lock, I updated my solution to use 'postfix status' as my solution for our servers in place of kill -0.... 1 line fix.

Note: /opt/zimbra/common/sbin/postfix  calls /opt/zimbra/common/libexec/postfix-script where you can see start|stop|status and see how master determines it is running. If you look at the source for master.c you can see how the -t option (testing for that lock) on master.pid

Reply:
I tend to agree with @JimDunphy here.  Postfix [master](https://linux.die.net/man/8/master) `-t` is designed to do its own sanity check:

> -t
> Test mode. Return a zero exit status when the master.pid lock file does not exist or when that file is not locked. This is evidence that the master(8) daemon is not running.

Maybe we could/should rely on the tool postfix provides in this case?

Reply:
It feels like @JimDunphy and @plobbes didn't even try their suggestion, because Zimbra delivers a Postfix version where `master -t` returns a non-zero exit status independent of the actual master daemon status:

```
[root@zimbra ~]# runuser -l zimbra -c 'zmcontrol -v'
Release 8.8.11_GA_3737.RHEL7_64_20181207111719 RHEL7_64 NETWORK edition, Patch 8.8.11_P2.
[root@zimbra ~]# ps aux | grep /opt/zimbra/common/libexec/master
root      7832  0.0  0.0  49792  1832 ?        Ss   13:43   0:00 /opt/zimbra/common/libexec/master -w
root     11647  0.0  0.0 112720   984 pts/0    S+   13:45   0:00 grep --color=auto /opt/zimbra/common/libexec/master
[root@zimbra ~]# cat /opt/zimbra/data/postfix/spool/pid/master.pid 
                            7832
[root@zimbra ~]# /opt/zimbra/common/libexec/master -t; echo $?
1
[root@zimbra ~]# runuser -l zimbra -c 'zmmtactl stop'
/postfix-script: stopping the Postfix mail system
Stopping saslauthd...done.
[root@zimbra ~]# ps aux | grep /opt/zimbra/common/libexec/master
root     12610  0.0  0.0 112720   984 pts/0    S+   13:47   0:00 grep --color=auto /opt/zimbra/common/libexec/master
[root@zimbra ~]# cat /opt/zimbra/data/postfix/spool/pid/master.pid 
                            7832
[root@zimbra ~]# /opt/zimbra/common/libexec/master -t; echo $?
1
[root@zimbra ~]# 
```

And just to make it even more clear:

```
[root@zimbra ~]# ps aux | grep /opt/zimbra/common/libexec/master
root     12610  0.0  0.0 112720   984 pts/0    S+   13:47   0:00 grep --color=auto /opt/zimbra/common/libexec/master
[root@zimbra ~]# rm -f /opt/zimbra/data/postfix/spool/pid/master.pid 
[root@zimbra ~]# cat /opt/zimbra/data/postfix/spool/pid/master.pid
cat: /opt/zimbra/data/postfix/spool/pid/master.pid: No such file or directory
[root@zimbra ~]# /opt/zimbra/common/libexec/master -t; echo $?
1
[root@zimbra ~]# 
```

And to ensure that `/opt/zimbra/data/postfix/spool/pid/master.pid` is really the PID file being handled by `master -t`:

```
[root@zimbra ~]# strace -f -- /opt/zimbra/common/libexec/master -t 2>&1 | tail
getuid()                                = 0
getgid()                                = 0
getegid()                               = 0
socket(AF_INET, SOCK_STREAM, IPPROTO_IP) = 9
close(9)                                = 0
getrlimit(RLIMIT_FSIZE, {rlim_cur=RLIM64_INFINITY, rlim_max=RLIM64_INFINITY}) = 0
chdir("/opt/zimbra/data/postfix/spool") = 0
access("pid/master.pid", F_OK)          = -1 ENOENT (No such file or directory)
exit_group(0)                           = ?
+++ exited with 0 +++
[root@zimbra ~]# 
```

Thus from my point of view, my pull request still remains the only working suggestion.

Reply:
I agree, I want this fixed also. I have a bug report ZBUG-798 from Dec 2018 that is still open.

The postfix start is guarded by this construct in **/opt/zimbra/common/libexec/postfix-script** which is what postfix calls their "management script" in the source code. 

start)

        $daemon_directory/master -t 2>/dev/null || {
                $FATAL the Postfix mail system is already running
                exit 1
        }

They have some additional checks before calling it directly like you have shown which changes the results as verified below. If you remove another processes files than odd things will happen. To make it clear:

[root@mail] ps aux | grep /opt/zimbra/common/libexec/master
root     **21527**  0.0  0.0  50816  3560 ?        Ss   Mar19   0:01 /opt/zimbra/common/libexec/master -w
[root@mail ~]# cat /opt/zimbra/data/postfix/spool/pid/master.pid 
                           **21527**
[root@mail ~]# /opt/zimbra/common/sbin/postfix  status
/postfix-script: the Postfix mail system is running: PID: **21527**
[root@mail ~]# /opt/zimbra/common/sbin/postfix stop
/postfix-script: stopping the Postfix mail system
/opt/zimbra/common/sbin/postfix  status
/postfix-script: the Postfix mail system is not running

[root@mail ~]# ps aux | grep /opt/zimbra/common/libexec/master
root     16293  0.0  0.0 103384  2188 pts/0    S+   07:39   0:00 grep /opt/zimbra/common/libexec/master

Given that postfix eventually needs to be started, it seemed like the simplest method was to use what was already there. That master -t logic is there if we like it or not and it appears to work on my instances and as I have shown above.  You still have to navigate that master -t logic on the start nonetheless. Given that they were already calling /opt/zimbra/common/sbin/postfix to start postfix, it seemed like calling it also for status in front of that logic was a prudent thing to do as that pathway appears to be working and tested.

Again... I wasn't recommending that master -t as you have shown but /opt/zimbra/common/sbin/postfix status... That is a second working solution in addition to your solution. 

I agree with Robert that it is time to close this bug. Sorry I commented now because the purpose wasn't to slow down a solution but show that it was trivial to fix. Unfortunately, there are more than a few other kill -0 patterns around but starting with the MTA is critical.

Reply:
I would also like to add that if people are removing the master pid file as a workaround to the kill -0 pattern they are setting themselves up for a bad day. That file is used for advisory locking so be extra careful when removing any file that another program uses. Also, the only reason to know about the pid file by name is because of the current kill -0 pattern or how Robert is discovering the master isn't running. Note: there are checks for multiple instances in that postfix-script in addition to some convenient functions. 

Again **I am not recommending anyone use master -t** but mentioned it show that postfix does test the lock via its management script that we already trust for start/stop in zimbra. It's only the status that is broken which fails to call postfix start because of pid wrap-around. I know the postfix forums are filled with people blowing away the pid file and then adding bug reports. :-) 

This is super easy to fix so why not in the next patch release?   Use the included postfix status call or the way @robert-scheck  has shown. They both work. The kill -0 pattern doesn't work.

Reply:
@robert-scheck believe it or not, but I tested on RHEL6 before (and again a few minutes ago) and found master -t working as expected.  Granted if it's not working across all os/platforms it's not a good solution.  So someone will have to make a call on this.

Reply:
Given the way community contributions are treated by Zimbra, it feels like Zimbra has absolutely no interest in them. So, will somebody make a decision to get this bug addressed (and when)?

Reply:
I generally agree that Zimbra doesn't have the best community involvement but in this case I think the original requests for changes were fine. Contributing to Open Source doesn't mean that all patches are applied without changes, that's what the PRs are for. And using the means provided by the subsystems sounds like a sensible approach.

I still think your approach using `/proc` would be fine as well BTW; theoretical portability is fine but in case of Zimbra it would help a lot if all the old bitrotten portability things would be ripped out which would increase readability a lot. I also agree that systemd would be a good approach theoretically but that would cause issues with the list of supported OSes.

So the issue is that `postfix status` works on Ubuntu (I did some quick tests on 14.04) and RHEL6 but not RHEL7? Or is it only `master -t` which seems not to work but `postfix status` maybe returns the proper results due to the additonal checks pointed out by @JimDunphy ?

Reply:
I'm sorry, but you unfortunately did not get the point of my complaint. This is not about that exactly my patch must be merged, but only about getting the bug itself fixed.

When I encountered this bug for the first time in May 2018, I reported it (as a paying customer) to the Zimbra support. I guess I shouldn't go into more details publically, but as of March 2019 the bug still exists even a "queasy" patch was provided in August 2018. Over the whole time frame, I did not see serious efforts by Zimbra to get this bug addressed using own developers (alternatively to my contribution).

RHEL/CentOS 6 reaches EOL on 2020-11-30, Ubuntu 14.04 LTS reaches EOL on 2019-04-25, RHEL/CentOS 7 and Ubuntu 16.04 LTS are supporting systemd, thus your expected issues with supported OSes are gone quite soon.

To continue more on-topic:

```
[root@zimbra ~]# runuser -l zimbra -c 'zmcontrol -v'
Release 8.8.11_GA_3737.RHEL7_64_20181207111719 RHEL7_64 NETWORK edition, Patch 8.8.11_P2.
[root@zimbra ~]# /opt/zimbra/common/sbin/postfix status
/postfix-script: the Postfix mail system is running: PID: 9854
[root@zimbra ~]# runuser -l zimbra -c 'zmmtactl stop'
/postfix-script: stopping the Postfix mail system
Stopping saslauthd...done.
[root@zimbra ~]# /opt/zimbra/common/sbin/postfix status
/postfix-script: the Postfix mail system is not running
[root@zimbra ~]# echo 1 > /opt/zimbra/data/postfix/spool/pid/master.pid 
[root@zimbra ~]# /opt/zimbra/common/sbin/postfix status
/postfix-script: the Postfix mail system is not running
[root@zimbra ~]# 
```